### ISSUE TITLE
Variable undefined check should goes through functions in front of them.

### ISSUE THREAD

**Issue by RedTailBullet** [Microsoft/TypeScript#25856]
## Search Terms
variable, check, undefined

## Suggestion
While checking whether a variable is undefined or null, checker should go through all the code including the function runned before the variable is used. 

## Use Cases
I have a function that required a non-undefined argument. I do the required field check in a separate function and if the required variable is undefined or null, it should throw an error and the variable is not really used.

However, the checker now is still showing me the error that this variable might be undefined.

## Examples
```
class example {
  private property: string

  public assignVariable (variable: string | undefined) {
    this.checkRequiredVariable(variable)
    this.property = variable // This is throwing the "Type 'undefined' is not assignable to type 'string'." error.
  }

  private checkRequiredVariable (variable: string | undefined) {
    if (variable == undefined)
      throw new Error('Variable is undefined or null')
  }
}
```  

PS: This code is just for the example, I know that this will throw a "No initialization" error as well.  

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by DanielRosenwasser** [comment]
Sounds like you want #17760; is that accurate?

**Comment by RedTailBullet** [comment]
@DanielRosenwasser He does mention this kind of thing but mine is not that complicated. If you think it's a duplication I'll accept that. Sorry for the duplicated issue.

**Comment by andy-ms** [comment]
To get type-checking currently, we use a pattern like this:
```ts
public assignVariable(variable: string | undefined) {
	this.property = this.checkRequiredVariable(variable)
}

private checkRequiredVariable(variable: string | undefined): string {
	if (variable == undefined)
		throw new Error('Variable is undefined or null')
	return variable
}
```

You can also make this generic:
```ts
function assertDefined<T>(x: T | null | undefined): T {
	if (x == undefined)
		throw new Error("Variable is undefined or null")
	return x
}
```
